src/crimson/os/seastore/segment_cleaner.h

```diff
@@ -18,15 +18,66 @@
 
 namespace crimson::os::seastore {
 
+struct segment_info_t;
+using unavail_segment_info_set_t =
+  std::list<segment_info_t*>;
+using unavail_segment_info_iterator =
+  unavail_segment_info_set_t::iterator;
+
+// position within segment_info_set_t
+using segment_pos_t = uint64_t;
+
+template <typename SegmentInfoT>
+struct segment_manager_info_t {
+  segment_manager_info_t() = default;
+  segment_manager_info_t(
+    device_id_t device_id,
+    device_segment_id_t num_segments,
+    segment_off_t segment_size,
+    segment_off_t block_size,
+    size_t empty_segments,
+    size_t size)
+    : device_id(device_id),
+      num_segments(num_segments),
+      segment_size(segment_size),
+      block_size(block_size),
+      empty_segments(empty_segments),
+      size(size)
+  {}
+
+  void init_segment_infos(SegmentInfoT&&);
+
+  unavail_segment_info_set_t usis;
```

Comment:
This is a secondary index of the contents of segment_infos containing those segments which are unavailable.  You really need to include a comment with an explanation for members like that, every time.

The only user I see for this is set get_available_bytes.  In that method, I see that we consider the portions of "unavailable" segments past the write pointer to be available.

mark_closed() calls mark_unavail() if we are still initting (!init_complete).  set_open() calls mark_unavail() unconditionally.

I don't understand what the intention here is.  The only case where space past the write_pointer should be considered "available" is if the segment is still open.  I think SegmentCleaner as a whole should simply track the set of open Segments along with their written_to values -- the SegmentProvider interface gets notified when segments are opened, closed, or written to.  written_to doesn't seem to make sense otherwise -- it's irrelevant for a closed segment and 0 for an empty one.

Reply:
I changed to track the available space of each segment manager using a counter, please take a look again.